File: Agent/.NET 6.0/Identity.cs
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Net;
using System.Security.Claims;

namespace Technologai
{
    public class Identity
    {
        public string? Name { get; private set; }
        public string AgentId { get; private set; }        
        public string? AgencyId { get; private set; }
        internal Authority Authority { get; private set; }
        internal string InstanceId { get; private set; }
        internal string InstanceSecret { get; private set; }        
        
        internal Dictionary<string, string> Tokens = new Dictionary<string, string>();
        internal string PublishMask => $"{AgencyId}/+";
        internal string SubscribeMemberMask => $"{AgencyId}/{AgentId}";
        internal string SubscribeAgencyMask => $"{AgencyId}/0";        

        public Identity(string authority, string instanceId, string instanceSecret, string agentId)
        {
            Authority = new Authority(authority);
            InstanceId = instanceId;
            InstanceSecret = instanceSecret;
            AgentId = agentId;
        }

        internal async Task Authenticate(string audience, string version = "1.0")
        {
            using (var httpClient = new HttpClient())
            {
                httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", Base64UrlEncoder.Encode($"{InstanceId}:{InstanceSecret}"));
                httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

                var parameters = new Dictionary<string, string>();
                parameters.Add("grant_type", "client_credentials");                
                parameters.Add("audience", audience);
                parameters.Add("version", version);
                parameters.Add("scope", $"agent_id:{AgentId}");

                var endpoint = Authority?.AuthUri + Authority?.TokenApi;

                var httpResponse = await httpClient.PostAsJsonAsync(endpoint, parameters);

                if (httpResponse.StatusCode == HttpStatusCode.OK)
                {
                    var tokenResponse = await httpResponse.Content.ReadFromJsonAsync<TokenResponse>();

                    if (tokenResponse != null)
                    {
                        foreach (Claim claim in new JwtSecurityTokenHandler().ReadJwtToken(tokenResponse.access_token).Claims)
                        {
                            if (claim.Type == "agency_id")
                            {
                                AgencyId = claim.Value;
                            }
                            if (claim.Type == "name")
                            {
                                Name = claim.Value;
                            }
                            if (claim.Type == "aud")
                            {
                                Tokens[claim.Value] = tokenResponse.access_token;
                            }
                        }
                        return;
                    }
                }
                throw new HttpRequestException("Unauthorized", null, httpResponse.StatusCode);
            }
        }
        public string GetMaskedTopic(string topic)
        {
            string[] topicParts = topic.Split('/');
            string[] maskParts = PublishMask.Split('/');

            if (topicParts.Length != 2)
            {
                throw new ArgumentException(nameof(topic));
            }

            for (int i = 0; i < topicParts.Length; i++)
            {
                topicParts[i] = maskParts[i] == "+" ? topicParts[i] : maskParts[i];
            }
            return string.Join('/', topicParts);
        }

        internal class TokenResponse
        {
            public string? access_token { get; set; }
            public string? token_type { get; set; }
            public int? expires_in { get; set; }
        }
    }
}

File: Agent/python_agent/identity.py
import base64
import requests
import jwt

from constants import TOKEN_API

class Identity:
    def __init__(self, auth_uri, client_id, client_secret, member_id):
        self.auth_uri = auth_uri
        self.client_id = client_id
        self.client_secret = client_secret
        self.member_id = member_id
        self.agency_id = None
        self.tokens = {}

    @property
    def name(self):
        if self.id_token:
            return self.id_token.get('name')
        return None

    @property
    def id(self):
        if self.id_token:
            return self.id_token.get('sub')
        return None

    @property
    def subscribe_member_mask(self):
        return f"{self.agency_id}/{self.member_id}"

    @property
    def subscribe_agency_mask(self):
        return f"{self.agency_id}/0"

    async def authenticate(self, audience):
        headers = {
            'Authorization': 'Bearer ' + base64.b64encode(f"{self.client_id}:{self.client_secret}".encode()).decode(),
            'Accept': 'application/json'
        }

        parameters = {
            'grant_type': 'client_credentials',
            'scope': f'member:{self.id}',
            'audience': audience
        }

        endpoint = self.auth_uri + TOKEN_API
        response = requests.post(endpoint, headers=headers, json=parameters)

        if response.status_code == 200:
            token_response = response.json()

            if token_response is not None:
                decoded_token = jwt.decode(token_response['access_token'], options={"verify_signature": False})  # TODO verify signature

                for claim in decoded_token:
                    if claim == 'agency_id':
                        self.agency_id = decoded_token[claim]
                    if claim == 'name':
                        self.name = decoded_token[claim]
                    if claim == 'aud':
                        self.tokens[decoded_token[claim]] = token_response['access_token']
                return

        raise Exception('Unauthorized')
